### PR TITLE
Adding interop tests

### LINKED ISSUE #20685: Add interoperability tests with GnuTLS and NSS
Hi,

We would like to add TLS interoperability tests into OpenSSL upstream test suite. The motivation is to find problems sooner in the development process and therefore increase software quality.

In the past, we did something similar in the GnuTLS project (https://gitlab.com/gnutls/gnutls/-/pipelines/828498691/ - “test”/”fedora-interop/test”, basically testing TLS interoperability of GnuTLS changes with released OpenSSL and NSS).

Would you be open to including similar testing to OpenSSL? If so, are there some rules we should be aware of? Can we discuss the technical approach as part of this issue, or is there a better way?

Thanks, 
Peter

### PR DESCRIPTION
**Pull request by nhorman**
Fedora has some fairly nice interoperability tests that we can leverage  to build a PR and test it against gnutls and nss libraries.  This PR adds the interop-tests.yml ci job to do that work, and runs the interop tests from beaker.

Fixes #20685

### CODE REVIEW

File: .github/workflows/interop-tests.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name: Interoperability tests with GnuTLS and NSS
+on: [pull_request, push]
```
Comment (t8m) [inline .github/workflows/interop-tests.yml:2]: Given these are very unlikely to fail they should run on push only. But to manifest they are working there should be a separate commit marked with `drop!` to run them on pull request in this PR.

Comment (nhorman) [reply .github/workflows/interop-tests.yml:2]: oh, yeah, I've only got that now to test out the containerized build (the build container needs some work to build openssl from the target branch, rather than having it be built once when the container is built)

File: .github/workflows/fedora-interop.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM fedora:38
+RUN dnf -y update
+RUN dnf -y install perl
+RUN dnf -y install git tmt-all
+RUN dnf -y remove openssl
+RUN cd / && git clone https://gitlab.com/not4pedro/interop
```
Comment (ep69) [inline .github/workflows/fedora-interop.Dockerfile:6]: Please use `https://gitlab.com/redhat-crypto/tests/interop`

File: .github/workflows/fedora-interop.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+RUN cd interop && git checkout 68fb662546d0b1f1fe5cab88218524e7db5764d3
```
Comment (ep69) [inline .github/workflows/fedora-interop.Dockerfile:7]: Doable and probably ok as the test repo is not really large. I can also create a specific branch for openssl if you like.

Other approach, that we used in GnuTLS upstream, was having tests directly in the GnuTLS repo as a git submodule.

Whatever works best for openssl.

File: .github/workflows/interop-tests.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-22.04
+    timeout-minutes: 90
+    strategy:
+      fail-fast: false
+      matrix:
+        COMPONENT: [openssl-upstream-gnutls, openssl-upstream-nss]
```
Comment (ep69) [inline .github/workflows/interop-tests.yml:10]: We can do even more parallelization if desired, can get the overall time to under 30 minutes. For inspiration, GnuTLS config: https://gitlab.com/gnutls/gnutls/-/blob/master/.gitlab-ci.yml?ref_type=heads#L337

File: .github/workflows/interop-tests.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    env:
+      COMPONENT: ${{ matrix.COMPONENT }}
+    # Code below inspired by https://fedoramagazine.org/github-actions-use-podman-to-run-fedora-linux/
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+      - name : Run tests
+        run: |
+          docker run --sysctl net.ipv6.conf.lo.disable_ipv6=0 -di -v /home/runner/work/openssl/openssl:/openssl --name fedora38 docker.io/nhorman/fedora38-interop:latest
```
Comment (ep69) [inline .github/workflows/interop-tests.yml:20]: As you mentioned, probably not ideal to use personal docker.io account.

How do we plan to maintain the container in future?

**Review by ep69 — COMMENTED** [review]
I don't see the interoperability tests running in "Checks" tab of the pull request, is it expected? Did I overlook something?

File: .github/workflows/interop-tests.yml (hunk 5 above)
Comment (quarckster) [reply .github/workflows/interop-tests.yml:20]: why would not use preinstalled docker?

```yaml
name: Interoperability tests with GnuTLS and NSS
on: [pull_request, push]
jobs:
  test:
    runs-on: ubuntu-22.04
    timeout-minutes: 90
    strategy:
      fail-fast: false
      matrix:
        COMPONENT: [openssl-upstream-gnutls, openssl-upstream-nss]
    env:
      COMPONENT: ${{ matrix.COMPONENT }}
    container: docker.io/nhorman/fedora38-interop:latest 
    steps:
      - uses: actions/checkout@v4
        with:
          repo-token: ${{ secrets.GITHUB_TOKEN }}
      - name : Run tests
        run: |
          ln -s /interop . && ls -l ./interop/ && ./config --banner=Configured no-makedepend enable-fips --strict-warnings && perl configdata.pm --dump
          make -s -j4
          echo "Discovering tests"
          tmt run plans -n $COMPONENT discover -v
          echo "Running tests"
          tmt run -a plans -n $COMPONENT provision -h local execute -h tmt --interactive
          echo "Finished - important to prevent unwanted output truncating"
```

File: .github/workflows/fedora-interop.Dockerfile (hunk 2 above)
Comment (quarckster) [reply .github/workflows/fedora-interop.Dockerfile:6]: I don't see the point of having a custom container. `tmt` provides its own on `quay.io/testing-farm/tmt-all`.

Comment (nhorman) [reply .github/workflows/fedora-interop.Dockerfile:6]: @quarckster I didn't realize there was one.  If its capable of also doing builds, yeah, I'll switch to that

File: .github/workflows/interop-tests.yml (hunk 5 above)
Comment (nhorman) [reply .github/workflows/interop-tests.yml:20]: @quarckster If you look at the subsequent commits, thats been removed

Comment (quarckster) [reply .github/workflows/interop-tests.yml:20]: You don't have to execute `docker` command. If you specify `container` on a job the context will be inside the container.

File: .github/workflows/fedora-interop.Dockerfile (hunk 2 above)
Comment (nhorman) [reply .github/workflows/fedora-interop.Dockerfile:6]: @quarckster the tmt-all container doesn't support gcc at the moment.  It can be added at run time of course, but that sort of makes it a 6 of one / 0.5 of a dozen of the other situation.  Either we modify an existing container, or we build our own, Not sure which is preferable, but I would lean towards building our own container, as it gives us the possibility to fix bugs in the container more easily

**Comment by nhorman** [comment]
@ep69 to your comment [here](https://github.com/openssl/openssl/pull/22726#pullrequestreview-1731805792)

The check is there, you may have just missed it.  Heres the link:
https://github.com/openssl/openssl/actions/runs/6868719823/job/18679926043?pr=22726

File: .github/workflows/fedora-interop.Dockerfile (hunk 3 above)
Comment (nhorman) [reply .github/workflows/fedora-interop.Dockerfile:7]: Yeah, I'm trying to decide what would be best here.  I like the idea of having the interop tests be present in the openssl repo so we have a bit more of a direct path to address fixes in openssl directly, but if these tests are under active development I don't think that really makes sense, as I'd like to incorporate new tests as they are developed.

does the interop repo have plans for tagging stable points or stable branches?  If so, I think the best approach might be to not incorporate the tests in the docker image, but rather clone them at run time in the ci yaml

**Comment by nhorman** [comment]
@ep69 to your comment [here](https://github.com/openssl/openssl/pull/22726#discussion_r1394069069), the answer is that we are currently unsure.  There are a few options:
1) increase the size of the public openssl repo container storage and use that
2) host the containers on the openssl enterprise storage
3) use an alternate public repository with openssl enterprise credentials
4) something we haven't thought of yet

Right now the personal account on docker.io that I'm using is strictly just temporary to be able to test this out, that will change with whatever decision is made above prior to this getting merged

File: .github/workflows/interop-tests.yml (hunk 5 above)
Comment (nhorman) [reply .github/workflows/interop-tests.yml]: ack I've updated the yaml to use the syntax you suggested, that looks much nicer!

File: .github/workflows/interop-tests.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      fail-fast: false
+      matrix:
+        COMPONENT: [openssl-upstream-gnutls, openssl-upstream-nss]
+    env:
+      COMPONENT: ${{ matrix.COMPONENT }}
+    # Code below inspired by https://fedoramagazine.org/github-actions-use-podman-to-run-fedora-linux/
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (quarckster) [inline .github/workflows/interop-tests.yml:22]: I don't think this is necessary. First of all, there is no such parameter `repo-token` https://github.com/actions/checkout#usage. but there is `token` and it has the current repo token as a default value.

File: .github/workflows/interop-tests.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+name: Interoperability tests with GnuTLS and NSS
+on: [pull_request, push]
+jobs:
+  test:
+    runs-on: ubuntu-22.04
+    container:
+      image: docker.io/nhorman/fedora38-interop:latest
+      options: --sysctl net.ipv6.conf.lo.disable_ipv6=0
+      volumes:
+        - /home/runner/work/openssl/openssl:/openssl
```
Comment (quarckster) [inline .github/workflows/interop-tests.yml:10]: This might be not needed as well. `actions/checkout@v4` is executed within the container.

Comment (nhorman) [reply .github/workflows/interop-tests.yml]: yup, you're right, changing to the container syntax you referenced lets us do the checkout inside the container, so volume mounts aren't needed

**Comment by nhorman** [comment]
@ep69 Noticed that the interop tests from that repo don't actually test an installed version.  They contain beaker scripts which install the openssl rpm built from whatever version the fedora repo has published.  While thats useful in Fedora testing, its not as useful here, where the goal is to test whatever version of openssl is built and installed at the time (the ci built version).  Is it possible to augment the interop tests to skip the installation of openssl in the various runtests.sh scripts?

**Comment by Cris269** [comment]
Enviado desde mi smartphone Samsung Galaxy.
-------- Mensaje original --------De: Neil Horman ***@***.***> Fecha: 14/11/23  05:04  (GMT-06:00) A: openssl/openssl ***@***.***> CC: Subscribed ***@***.***> Asunto: [openssl/openssl] Adding interop tests (PR #22726) From #22043

Checklist


 documentation is added or updated
 tests are added or updated




You can view, comment on, or merge this pull request online at:
  https://github.com/openssl/openssl/pull/22726

Commit Summary

  222da66  Adding interop tests


File Changes  (1 file)

  
    A
    .github/workflows/interop-tests.yml
    (36)
  


Patch Links:

  https://github.com/openssl/openssl/pull/22726.patch
  https://github.com/openssl/openssl/pull/22726.diff


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/openssl/openssl/pull/22726",
"url": "https://github.com/openssl/openssl/pull/22726",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by ep69** [comment]
> @ep69 Noticed that the interop tests from that repo don't actually test an installed version. They contain beaker scripts which install the openssl rpm built from whatever version the fedora repo has published. While thats useful in Fedora testing, its not as useful here, where the goal is to test whatever version of openssl is built and installed at the time (the ci built version). Is it possible to augment the interop tests to skip the installation of openssl in the various runtests.sh scripts?

That's interesting. From the quick look, I don't seem to be able to identify the change that causes this. Perhaps @not4Pedro know? I'll look into it deeper if not.

**Comment by ep69** [comment]
@nhorman ah, I got an impression that you are saying that the tests from `not4pedro` repositories are fine and problem is only in the `redhat-crypto/tests/interop`. That's not the case, right?

I'll investigate how to best avoid it.

**Comment by nhorman** [comment]
@ep69 no, apologies.  the issues manifests in both repositories.  

I noticed it when I observed that we are building openssl in a directory that isn't in PATH, but the tests were still running fine when executing 'openssl'.  Examining the container when run locally I observed that, regardless of the container running
```
dnf -y remove openssl
```
at build time, after running the interop tests manually /usr/bin/openssl exists again

If I remember the beaker library properly, its happening because all the runtests.sh scripts include openssl in the PACKAGE[S] environment variable, and when rlAssertInstalled (I think thats the beaker function), it just re-installs openssl.

I think what we need is one of the following:
1) a removal of openssl from the packages env variable, allowing it to assume that openssl is installed already, combined with a running of make install during the openssl build in ci
2) a method to override the PACKAGE variable during the running of the interop tests, combined with a augmenting of PATH and LD_LIBRARY_PATH to point to the locally built openssl instance

I think (1) is probably not overly workable as removing openssl from PACKAGES is going to break fedora testing I'm sure.

(2) seems like a reasonable option I think
2)

**Comment by nhorman** [comment]
ping, @ep69 any thoughts on how to manage the beaker tests in an upstream ci context?

**Comment by nhorman** [comment]
Container storage is handled I believe.  I need to remember what @quarckster setup for us in terms of container storage, but as long as we can control the contents of the container, storing it shouldn't be an issue.

As far as holding snapshot of the interop tests internal to the container (rather than pulling it every time), thats a-ok with me, I would just ask that we have some way to update when needed (i.e. the redhat repo would have to have some branch/mechanism to update the variant of the beaker tests that allow for pre-installed openssl binaries, rather than rpm-installed ones).

**Comment by ep69** [comment]
I believe I am pretty close to reasonable solution, still looking into this. Might have something today or (hopefully) latest later this week :muscle:

**Comment by nhorman** [comment]
@ep69 thanks, and no rush, just let me know when you have it all ready

**Comment by quarckster** [comment]
I don't see the point in the Dockerfile here. These commands should be executed in a regular GH Actions workflow

**Comment by quarckster** [comment]
@nhorman that's right. `registry.fedoraproject.org/fedora:39` should be fine.

**Comment by nhorman** [comment]
@quarckster Ok, I can modify the ci job to do it that way, no problem

**Comment by nhorman** [comment]
@ep69 In the current branch setup, I seem to be seeing an issue in which beakerlib is attempting to start a new VM to run gnutls, but I can't see where in the interop test plans actually do that provisoning.  Thats going to be a problem, as we can't spawn a new vm from within a container.  And oddly I don't recall that happening before.  Do you have any insight as to where that provisioning is happening?  We're going to need to run gnutls/nss from within the same container we are testing from.

**Comment by nhorman** [comment]
thank you for the notes @ep69 , I've got a few other things on my plate at the moment, but will give your suggestions a try in the am

**Comment by nhorman** [comment]
Ok, I think I've got it working, and the rpmbuild step was less worrisome than I thought it would be.  @ep69 @quarckster please take a look, and let me know what you think.  If you're all comfortable with it, I'll take this out of draft for a proper review.

**Comment by nhorman** [comment]
@ep69  please note this stanza in the CI run output:
```
 prepare task #2: recommends on default-0
        how: install
        summary: Install recommended packages
        name: recommends
        package: openssl, crypto-policies-scripts, net-tools and 4 more
```

I can't quite tell if its trying to re-install openssl (it shouldn't given that the locally built rpm is at a later semantic version, but I would appreciate it if you could confirm

File: .github/workflows/fedora-interop.Dockerfile [hunk 8]
```diff
@@ -0,0 +1,8 @@
+FROM fedora:38
+RUN dnf -y update
+RUN dnf -y install perl
+RUN dnf -y install git tmt-all
+RUN dnf -y remove openssl
+RUN cd / && git clone https://gitlab.com/redhat-crypto/tests/interop
+RUN cd interop && git checkout 3def3aa0ef173b1479611909f13ac4abb604e62b
+VOLUME /openssl
```
Comment (quarckster) [inline .github/workflows/fedora-interop.Dockerfile:8]: I guess this file is not needed anymore.

**Review by quarckster — COMMENTED** [review]
LGTM

File: .github/workflows/fedora-interop.Dockerfile (hunk 8 above)
Comment (nhorman) [reply .github/workflows/fedora-interop.Dockerfile:8]: yep, I'll remove it

**Comment by nhorman** [comment]
@ep69 scratch my previous comment, I managed to confirm the right version of openssl was run by dumping the openssl version at the end of the interop tests.  We're good there

File: .github/workflows/interop-tests.yml (hunk 4 above)
Comment (ep69) [reply .github/workflows/interop-tests.yml:13]: Let's simplify this:
```
        COMPONENT: [gnutls, nss]
```

File: .github/workflows/interop-tests.yml [hunk 9]
```diff
@@ -0,0 +1,44 @@
+name: Interoperability tests with GnuTLS and NSS
+on: [pull_request, push]
+jobs:
+  test:
+    runs-on: ubuntu-22.04
+    container:
+      image: docker.io/fedora:39
+      options: --sysctl net.ipv6.conf.lo.disable_ipv6=0
```
Comment (ep69) [inline .github/workflows/interop-tests.yml:8]: I don't know if nss tests fail with this. If so, I suggest removing.

File: .github/workflows/interop-tests.yml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        run: |
+          mkdir -p /build/SPECS && cd /build && echo -e "%_topdir /build\n%_lto_cflags %{nil}" >~/.rpmmacros && rpmdev-setuptree
+          cd /build && cp /__w/openssl/openssl/interop/openssl.spec SPECS/ && cd SPECS/ && \
+          source /__w/openssl/openssl/VERSION.dat && \
+          sed -i "s/^Version: .*\$/Version: $MAJOR.$MINOR.$PATCH/" openssl.spec && \
+          sed -i 's/^Release: .*$/Release: dev/' openssl.spec
+          yum-builddep -y /build/SPECS/openssl.spec # just for sure nothing is missing
+          mkdir -p /build/SOURCES
+          tar --transform "s/^__w\/openssl\/openssl/openssl-$MAJOR.$MINOR.$PATCH/" -czf /build/SOURCES/openssl-$MAJOR.$MINOR.$PATCH.tar.gz /__w/openssl/openssl/
+          rpmbuild -bb /build/SPECS/openssl.spec && dnf update -y /build/RPMS/x86_64/openssl-*
```
Comment (ep69) [inline .github/workflows/interop-tests.yml:37]: `dnf update` is probably redundant considering the next line with `dnf install`.

File: .github/workflows/interop-tests.yml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+          dnf install -y /build/RPMS/x86_64/openssl-*
+      - name: Run interop tests
+        run: |
+          cd interop
+          tmt run -a plans -n $COMPONENT provision -h local execute -h tmt --interactive
```
Comment (ep69) [inline .github/workflows/interop-tests.yml:42]: Please try it likethis:
```
tmt run -av plans -n interop tests -f "tag: interop-openssl & tag: interop-$COMPONENT" provision -h local execute -h tmt --interactive
```

**Review by ep69 — CHANGES_REQUESTED** [review]

**Comment by ep69** [comment]
@nhorman added suggestions, please have a look.

File: .github/workflows/interop-tests.yml (hunk 9 above)
Comment (nhorman) [reply .github/workflows/interop-tests.yml:8]: Its a bit counter-intuitive, but that line explicitly _enables_ ipv6, which I think is what you mentioned was required before, and so the nss and gnutls tests are passing currently.

Comment (ep69) [reply .github/workflows/interop-tests.yml:8]: Ah, yes, sorry for the noise :man_facepalming:

**Comment by nhorman** [comment]
@ep69 With this change made:
```
tmt run -av plans -n interop tests -f "tag: interop-openssl & tag: interop-$COMPONENT" provision -h local execute -h tmt --interactive
```
The tests fail:
https://github.com/openssl/openssl/actions/runs/7151945800/job/19476760560?pr=22726

Reverting it to the initial invocation passed previously

**Comment by nhorman** [comment]
Interesting, @ep69 with the adjustments as you suggested in the current branch, the interop tests run fine, but we actually hit a single failure:
```
fail /Interoperability/gnutls_TLSv1-2-with-OpenSSL
```

Looking at the logs it appears the gnutls test failed one instance when connecting over tls1.2.  It appears (if I'm reading it right) the gnutls client appears to have sent a client hello, but the server never received it.  Timing error in the test?  Legitimate interop issue?

**Comment by nhorman** [comment]
hmm, didn't happen again on rerun, transient failure I guess..

**Review by ep69 — APPROVED** [review]
Looks great to me.

@nhorman thank  you a lot for all the work you've put into this, I appreciate it a lot.

**Comment by nhorman** [comment]
@ep69 and you sir, thanks for offering this up to us, I think its a valuable test to add to our suite!

File: .github/workflows/interop-tests.yml (hunk 1 above)
Comment (quarckster) [reply .github/workflows/interop-tests.yml:2]: @nhorman do you plan to change triggers to `push` only as @t8m proposed?

Comment (nhorman) [reply .github/workflows/interop-tests.yml:2]: oh, yeah, forgot to reset that after I finished my testing, will fix

Comment (nhorman) [reply .github/workflows/interop-tests.yml]: fixed

**Review by quarckster — APPROVED** [review]
LGTM

**Review by t8m — CHANGES_REQUESTED** [review]
The test takes over one hour to run: https://github.com/openssl/openssl/actions/runs/7151995569/job/19476853336

Please change the scheduling to make it a daily job done at the same time as the other daily jobs run.

**Comment by ep69** [comment]
Weird. Last time I looked it took like 30 minutes :thinking:

**Comment by nhorman** [comment]
done

**Review by t8m — APPROVED** [review]

File: .github/workflows/interop-tests.yml [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      image: docker.io/fedora:39
+      options: --sysctl net.ipv6.conf.lo.disable_ipv6=0
+    timeout-minutes: 90
+    strategy:
+      fail-fast: false
+      matrix:
+        COMPONENT: [gnutls, nss]
+    env:
+      COMPONENT: ${{ matrix.COMPONENT }}
+    # Code below inspired by https://fedoramagazine.org/github-actions-use-podman-to-run-fedora-linux/
```
Comment (not4Pedro) [inline .github/workflows/interop-tests.yml:18]: Just looking at how much this setup has deviated from the initial one. You might want to remove this (inherited) reference as, at this point, it seems to me quite irrelevant/unhelpful so you can keep this code nice and clean.
Thank you everyone for all the work.

**Review by t8m — APPROVED** [review]

**Review by jamuir — COMMENTED** [review]
I am late to the conversation, so apologies for the silly questions :-)

Neil, you wrote:

> we need to be able to test against an arbitrary installed version of openssl ('arbitrary' implying whatever version the ci built and installed within the container environment

How do I map the version being considered by our CI to the contents of the yml file?  Are the sources for the version the CI wants to test magically present under the path `/__w/openssl/openssl` ?

Also, I see we are using the tag "39" to refer the container image (i.e. `docker.io/fedora:39`).  Would it be better to refer to the image using its sha256 digest?  Tags can get updated and we might unexpectedly pull down a different image.

**Comment by quarckster** [comment]
> Also, I see we are using the tag "39" to refer the container image (i.e. `docker.io/fedora:39`). Would it be better to refer to the image using its sha256 digest? Tags can get updated and we might unexpectedly pull down a different image.

Despite in general it's good practice to use the digest I don't think it's relevant to this case. I'm pretty sure that possible image updates will break anything. On the other hand, the digest will hide the Fedora version which we definitely want to know.

**Comment by nhorman** [comment]
@jamuir in answer to your questions:
1) yes, the path /__w/openssl is bind mounted from the CI jobs file system into the container - i.e. whatever PR/branch the CI job tests is whats getting built inside the container.  I certainly grant you that it is non-obvious (I had to go dig through the github docs to find what that path was.  I can add a comment in the yaml file if you think that will make it more clear

2) I think the use of the tag here is probably better (i.e. fedora:39 says to the user we're using a fedora 39 env).  A sha has would potentially be more stable (The fedora maintainers could certainly relabel stuff under the covers, but that would be a bad practice on their part I think for a number of reasons), but fedora:sha256@8987324678324837498237abcdef is not super clear to a user what we are using.  I had considered using fedora:latest instead, but I'm not sure getting the latest version at random times is really what we want, though opinions are welcome.

**Comment by jamuir** [comment]
> I can add a comment in the yaml file if you think that will make it more clear

That's a good idea.  That will save some time in the future when we look at this again after forgetting how it works :-)

**Comment by nhorman** [comment]
hey, at least theres a process :)

**Review by jamuir — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: .github/workflows/interop-tests.yml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      COMPONENT: ${{ matrix.COMPONENT }}
+    steps:
+      - uses: actions/checkout@v4
+      - name : Install needed tools
+        run: |
+          dnf -y install perl gcc rpmdevtools dnf-utils make tmt-all beakerlib \
+                 fips-mode-setup crypto-policies-scripts
+      - name: install interop tests
+        run: |
+          cd /__w/openssl/openssl
```
Comment (hlandau) [inline .github/workflows/interop-tests.yml:31]: Is this prefix not available as an environment variable? I assume GitHub reserves the right to change this.

**Review by hlandau — CHANGES_REQUESTED** [review]

File: .github/workflows/interop-tests.yml (hunk 13 above)
Comment (nhorman) [reply .github/workflows/interop-tests.yml:31]: Unfortunately no, and the docs don't mention anything about a workdir env variable.  You can see here:
https://github.com/openssl/openssl/actions/runs/7151995569/job/19476853336
That in the Initialize container step, github starts the container with the --workdir setting to set that path, but never sets a corresponding environment variable for us to fetch that value from.

We could manually set it and reference it, but that doesn't help if github changes that path.  I'm guessing they just haven't thought about the implications of doing so.

Only way I could see around this would be to explicitly add a bind mount in the volumes section, but that comes with its own issues, notably if github changes the runners physical path, it still breaks, and the fact that we would have two bind mounts to the physical path in the container, which, while allowed, would seem odd and prone to odd behaviors I think

Comment (hlandau) [reply .github/workflows/interop-tests.yml:31]: https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

What about `$GITHUB_WORKSPACE`?

Adding `export` to a workflow might be a useful aid to get a sense of what variables are available.

Comment (nhorman) [reply .github/workflows/interop-tests.yml]: oh, yes, that does map properly, I'll use that and add an export step to dump the environment

Comment (nhorman) [reply .github/workflows/interop-tests.yml]: Updated to use GITHUB_WORKSPACE

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you.
